### PR TITLE
Add DistilBertForCausalLM

### LINKED ISSUE #7397: Add DistilBERTGeneration comparable to BertGeneration
# 🚀 Feature request

I noticed the new `BertGeneration` class, which uses BERT-style models as both encoder and decoder, as well as the more general `EncoderDecoder` class. This is all great stuff! It would also be great to be able to use distilled models. I believe this is possible for the encoder, but for the decoder a language head must be added.

Since DistilBert is implemented as its own model, and not as a BertModel, I don't think it's possible (or at least it's not easy) for the end user to do this. At least not loading pretrained models, since any pretrained model needs to be a type approved by `AutoModelForCausalLM`.

## Motivation

Same motivation as using distilled models in general. Same results at higher speed, this time applied to an `EncoderDecoder` model.

## Your contribution

Happy to be an alpha tester for this feature

### PR DESCRIPTION
**Pull request by KMFODA**
# What does this PR do?

Similar to the `BertLMHeadModel` this PR aims to add a `DistilBertForCausalLM` model in `modeling_distilbert.py`.

Fixes #7397 
Replaces #8387

## Who can review?

@patil-suraj

### CODE REVIEW

**Comment by KMFODA** [comment]
Hey @patil-suraj, I've worked on your very helpful comments in #8387 and implemented them in this PR. Apart from a few formatting changes which I still need to work on the main test that is still failing is `python3 -m pytest tests/test_modeling_distilbert.py::DistilBertModelTest::test_for_causal_lm_decoder`. 

I'm still debugging this but it seems that the `cross_attention_outputs` aren't being produced when the model is run as a decoder because the `encoder_hidden_states` are empty. I'll try and find out why that's the case.

**Comment by KMFODA** [comment]
Hey @patil-suraj. Managed to fix the few tests that were failing and now `make tests` passes for all tests. I'm still seeing a few failed tests in the CI but the error messages don't seem to be related to the DistilBert model.

**Comment by patil-suraj** [comment]
Hi @KMFODA , yes the errors don't look related to this PR. Could you please rebase your branch and push again ? That should fix the CI issue. Let me know if it doesn't.

**Comment by KMFODA** [comment]
Done! Thanks for the suggestion @patil-suraj. Code now passes all checks. :)

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 1]
```diff
@@ -92,7 +93,11 @@ def __init__(self, config):
         self.LayerNorm = nn.LayerNorm(config.dim, eps=1e-12)
         self.dropout = nn.Dropout(config.dropout)
 
-    def forward(self, input_ids):
+        # position_ids (1, len position emb) is contiguous in memory and exported when serialized
+        self.register_buffer("position_ids", torch.arange(config.max_position_embeddings).expand((1, -1)))
+        self.position_embedding_type = getattr(config, "position_embedding_type", "absolute")
```
Comment (patil-suraj) [inline src/transformers/models/distilbert/modeling_distilbert.py:98]: If we add this then when loading distilbert model it will print warning  like this
`Some weights of DistilBertModel were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['distilbert.embeddings.position_ids']`.

Also, distilbert can use sinusoidal positional embedding which have no hard limit on maximum sequence length.

So I would remove this and keep the logic as it is for this model.

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 2]
```diff
@@ -145,7 +154,19 @@ def prune_heads(self, heads):
...
         self.pruned_heads = self.pruned_heads.union(heads)
 
-    def forward(self, query, key, value, mask, head_mask=None, output_attentions=False):
+    def forward(
+        self,
+        query,
+        key,
+        value,
+        mask,
+        attention_mask=None,
```
Comment (patil-suraj) [inline src/transformers/models/distilbert/modeling_distilbert.py:163]: mask and `attention_mask` seem to be doing the same thing, just renaming mask to `attention_mask` would be sufficient.

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 3]
```diff
@@ -174,15 +195,54 @@ def unshape(x):
             """ group heads """
             return x.transpose(1, 2).contiguous().view(bs, -1, self.n_heads * dim_per_head)
 
+        # q = shape(self.q_lin(query))  # (bs, n_heads, q_length, dim_per_head)
+        # k = shape(self.k_lin(key))  # (bs, n_heads, k_length, dim_per_head)
+        # v = shape(self.v_lin(value))  # (bs, n_heads, k_length, dim_per_head)
```
Comment (patil-suraj) [inline src/transformers/models/distilbert/modeling_distilbert.py:200]: This comment can be removed.

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 4]
```diff
@@ -145,7 +154,19 @@ def prune_heads(self, heads):
         self.dim = attention_head_size * self.n_heads
         self.pruned_heads = self.pruned_heads.union(heads)
 
-    def forward(self, query, key, value, mask, head_mask=None, output_attentions=False):
+    def forward(
+        self,
+        query,
+        key,
+        value,
```
Comment (patil-suraj) [inline src/transformers/models/distilbert/modeling_distilbert.py:161]: It would be nice if we could refactor this method to just pass `hidden_states` instead of `query, key, value` similar to `BertSelfAttention`. 

For self attention `hidden_states` are used as query, key, and value and for cross-attention encoder_hiddem_states` are used as key and value.

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 5]
```diff
@@ -174,15 +195,54 @@ def unshape(x):
...
+
+        # If this is instantiated as a cross-attention module, the keys
+        # and values come from an encoder; the attention mask needs to be
+        # such that the encoder's padding tokens are not attended to.
+        is_cross_attention = encoder_hidden_states is not None
+
+        if is_cross_attention and past_key_values is not None:
+            # reuse k,v, cross_attentions
+            k = past_key_values[0]
+            v = past_key_values[1]
```
Comment (patil-suraj) [inline src/transformers/models/distilbert/modeling_distilbert.py:210]: It would be nice to use `query, key, value` instead `q, k, v`.

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 6]
```diff
@@ -174,15 +195,54 @@ def unshape(x):
...
-        v = shape(self.v_lin(value))  # (bs, n_heads, k_length, dim_per_head)
 
         q = q / math.sqrt(dim_per_head)  # (bs, n_heads, q_length, dim_per_head)
         scores = torch.matmul(q, k.transpose(2, 3))  # (bs, n_heads, q_length, k_length)
         mask = (mask == 0).view(mask_reshp).expand_as(scores)  # (bs, n_heads, q_length, k_length)
         scores.masked_fill_(mask, -float("inf"))  # (bs, n_heads, q_length, k_length)
 
+        if attention_mask is not None:
+            # Apply the attention mask is (precomputed for all layers in DistilBertModel forward() function)
+            scores = scores + attention_mask
```
Comment (patil-suraj) [inline src/transformers/models/distilbert/modeling_distilbert.py:244]: as commented above `mask` and `attention_mask` are similar, so this can be removed.

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 7]
```diff
@@ -228,13 +293,27 @@ def __init__(self, config):
...
+            self.crossattention = MultiHeadSelfAttention(config)
         self.sa_layer_norm = nn.LayerNorm(normalized_shape=config.dim, eps=1e-12)
 
         self.ffn = FFN(config)
         self.output_layer_norm = nn.LayerNorm(normalized_shape=config.dim, eps=1e-12)
 
-    def forward(self, x, attn_mask=None, head_mask=None, output_attentions=False):
+    def forward(
+        self,
+        x,
```
Comment (patil-suraj) [inline src/transformers/models/distilbert/modeling_distilbert.py:309]:
```suggestion
        hidden_states,
```

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 8]
```diff
@@ -244,20 +323,62 @@ def forward(self, x, attn_mask=None, head_mask=None, output_attentions=False):
...
             output_attentions=output_attentions,
+            past_key_values=self_attn_past_key_value,
         )
+
         if output_attentions:
-            sa_output, sa_weights = sa_output  # (bs, seq_length, dim), (bs, n_heads, seq_length, seq_length)
+            (
+                sa_output,
+                sa_weights,
+            ) = self_attention_outputs  # (bs, seq_length, dim), (bs, n_heads, seq_length, seq_length)
```
Comment (patil-suraj) [inline src/transformers/models/distilbert/modeling_distilbert.py:344]: put the comment above this and make this a single line

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 9]
```diff
@@ -244,20 +323,62 @@ def forward(self, x, attn_mask=None, head_mask=None, output_attentions=False):
...
+            assert hasattr(
+                self, "crossattention"
+            ), f"If `encoder_hidden_states` are passed, {self} has to be instantiated with cross-attention layers by setting `config.add_cross_attention=True`"
+
+            # cross_attn cached key/values tuple is at positions 3,4 of past_key_value tuple
+            cross_attn_past_key_value = past_key_values[-2:] if past_key_values is not None else None
+            cross_attention_outputs = self.crossattention(
+                query=x,
+                key=x,
+                value=x,
```
Comment (patil-suraj) [inline src/transformers/models/distilbert/modeling_distilbert.py:367]: As commented above, would be better to refactor the attention to accept single `hidden_states` instead of `query, key, value`.

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 10]
```diff
@@ -273,13 +394,24 @@ def forward(self, x, attn_mask=None, head_mask=None, output_attentions=False):
...
         layer = TransformerBlock(config)
         self.layer = nn.ModuleList([copy.deepcopy(layer) for _ in range(config.n_layers)])
 
     def forward(
-        self, x, attn_mask=None, head_mask=None, output_attentions=False, output_hidden_states=False, return_dict=None
+        self,
+        x,
+        attn_mask=None,
+        head_mask=None,
+        position_ids=None,
```
Comment (patil-suraj) [inline src/transformers/models/distilbert/modeling_distilbert.py:408]: `position_ids` are not used, so we can remove this

**Review by patil-suraj — CHANGES_REQUESTED** [review]
Thanks a lot for working on this, the PR is in good shape now, great job!

I've left a few suggestions, let me know if something is not clear. Happy to help again :)

File: src/transformers/models/distilbert/modeling_distilbert.py (hunk 1 above)
Comment (KMFODA) [reply src/transformers/models/distilbert/modeling_distilbert.py:98]: Oh I see, but in that case how do we initialise self.position_ids which are used by line 112 below?

Comment (patil-suraj) [reply src/transformers/models/distilbert/modeling_distilbert.py:98]: distillbert computes the `position_ids` https://github.com/huggingface/transformers/blob/e783ea730456f67a246cb6d5db4893349f3e5af2/src/transformers/models/distilbert/modeling_distilbert.py#L104

**Comment by KMFODA** [comment]
Thanks for the review @patil-suraj. Made all the changes as per your comments.

**Review by patil-suraj — COMMENTED** [review]
HI @KMFODA 

Thanks a lot for improving this! It's looking good. 

We are still going to need a couple of revisions for this. I've left a  few comments, please let me know if you need any help with it.
Hope you don't mind 😄 .

**Comment by KMFODA** [comment]
No problem at all @patil-suraj. Once again I really appreciate the comments and help. Hopefully with this latest push your comments should be addressed :).

**Comment by KMFODA** [comment]
Hey @patil-suraj. Is there anything else I should change in the latest commit?

**Comment by patil-suraj** [comment]
Hey @KMFODA ! Sorry for being super slow here. Let me know if you still have time to work on this, I'll help with the merge conflicts and any other changes. Thanks :)

**Comment by KMFODA** [comment]
Hey @patil-suraj, no problems at all. Yes happy to work on it still. Do you see any conflicts that would not be resolved by a rebase?

**Comment by patil-suraj** [comment]
@KMFODA , Thanks a lot!
It would be awesome if you could rebase and resolve the conflicts. The PR is almost finished, will take another look once the conflicts are resolved. Let me know if you need any help with it!

**Comment by KMFODA** [comment]
Hi @patil-suraj , apologies for the delay. Was stuck in loops trying to ensure all tests passed and then rebasing. Should be good for review now.

**Comment by ArthurZucker** [comment]
Closing as this PR is super old and not needed anymore
